File: Dsa-day163.py
import heapq

class Solution:
    def mostBooked(self, n: int, meetings: list[list[int]]) -> int:
        meetings.sort()
        
        free_rooms = list(range(n))
        heapq.heapify(free_rooms)

        busy_rooms = []

        count = [0] * n

        for start, end in meetings:
            while busy_rooms and busy_rooms[0][0] <= start:
                _, room = heapq.heappop(busy_rooms)
                heapq.heappush(free_rooms, room)

            if free_rooms:
                room = heapq.heappop(free_rooms)
                count[room] += 1
                heapq.heappush(busy_rooms, (end, room))
            else:
                end_time, room = heapq.heappop(busy_rooms)
                duration = end - start
                new_end = end_time + duration
                count[room] += 1
                heapq.heappush(busy_rooms, (new_end, room))

        max_meetings = max(count)
        for i in range(n):
            if count[i] == max_meetings:
                return i

File: Dsa-day163.cpp
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
public:
    int mostBooked(int n, vector<vector<int>>& meetings) {
        sort(meetings.begin(), meetings.end());

        priority_queue<int, vector<int>, greater<int>> freeRooms;
        for (int i = 0; i < n; ++i)
            freeRooms.push(i);

        using pii = pair<long long, int>;  
        priority_queue<pii, vector<pii>, greater<pii>> busyRooms;

        vector<int> meetingCount(n, 0);

        for (auto& meeting : meetings) {
            int start = meeting[0];
            int end = meeting[1];

            while (!busyRooms.empty() && busyRooms.top().first <= start) {
                freeRooms.push(busyRooms.top().second);
                busyRooms.pop();
            }

            if (!freeRooms.empty()) {
                int room = freeRooms.top();
                freeRooms.pop();
                meetingCount[room]++;
                busyRooms.push({end, room});
            } else {
                auto [availableTime, room] = busyRooms.top();
                busyRooms.pop();
                long long duration = end - start;
                long long newEnd = availableTime + duration;
                meetingCount[room]++;
                busyRooms.push({newEnd, room});
            }
        }

        int maxMeetings = 0, resultRoom = 0;
        for (int i = 0; i < n; ++i) {
            if (meetingCount[i] > maxMeetings) {
                maxMeetings = meetingCount[i];
                resultRoom = i;
            }
        }

        return resultRoom;
    }
};
